docs/website/contents/explanations/design_goals.md

```diff
@@ -1 +1,63 @@
+The design should also support testing rare but expected scenarios (such as multiple slot leaders in Praos) by allowing overrides in protocol or ledger behavior at specific points.
+Furthermore, the system should facilitate the isolation testing of individual components.
+
+## Adaptability and Maintainability
+
+The consensus layer was developed as a replacement for a [previous implementation](https://github.com/input-output-hk/cardano-sl), with the immediate goal of transitioning from Byron/BFT to Shelley/Praos while supporting future ledger and protocol changes.
+This called for a flexible and adaptable design.
+Abstracting both the consensus algorithm and the ledger plays a crucial role in achieving this.
+Working with abstract interfaces prevents developers from making assumptions that may hold for one ledger but not others, avoiding costly fixes later.
+This abstraction also allows the consensus layer to be reused in other blockchain projects.
+Most importantly, an abstract design enables extensive testing with simpler mock ledgers, which are easier to set up and reason about compared to the complex real ledger.
+Abstraction is considered good engineering practice, enhancing clarity, reducing dependencies, and making the system easier to understand and maintain.
+
+## Composability
+
+Given the complexity and scale of the consensus layer codebase, it is essential to divide it into smaller, manageable components that can be understood and modified independently.
+Composability is a key technique employed to achieve this.
+A prime example is the Hard Fork Combinator (HFC), which enables the combination of different consensus protocols (such as BFT and Praos) and ledgers (such as Byron and Shelley) into a unified composite protocol or ledger for the hybrid chain.
+
+## Predictable Performance
+
+This goal ensures that node operators can configure nodes for "normal circumstances" without the network failing during infrequent but expected events.
+It aims to make node performance predictable, ensuring that the average-case scenario aligns with the worst-case scenario in terms of resource requirements—not only for security but also to maintain network stability with honest nodes.
+
+## Protection Against DoS Attacks
+
+The consensus layer must help safeguard the network against disruptions, making denial-of-service (DoS) attacks prohibitively expensive for adversaries.
+This involves design decisions that prevent attackers from easily causing a node to perform extensive, wasteful computations.
+For example, validating headers before downloading block bodies prevents attackers from forcing nodes to process potentially invalid blocks.
+The design often seeks a balance between the cost for an attacker to induce work and the cost for a node to defend against it.
+
+[^1]: See the IOSim monad and the `io-classes` package of [`io-sim`](https://github.com/input-output-hk/io-sim).
```

Comment:
```suggestion
[^1]: See the `IOSim` monad and the `io-classes` package of [`io-sim`](https://github.com/input-output-hk/io-sim).
```

Reply:
Possibly could be reformulated to point to https://input-output-hk.github.io/io-sim/